### PR TITLE
WIP `tools/terraform-generator`: fix issues in loadTestService

### LINKED ISSUE #2483: `tools/generator-terraform`: generate code of mapping sdk2schema will always set stat
in `func (m modelToModelAssignmentLine) assignmentForReadMapping` will init the sdk filed if it is nil in below code:

https://github.com/hashicorp/pandora/blob/345c443985de0cfdf8fc102023201a502d39964c/tools/generator-terraform/generator/mappings/assignment_model_to_model.go#L75-L83

, and this will always set an item to the model which will encoded to tf state file.

https://github.com/hashicorp/pandora/blob/c316d20771273b211c180c55e6dfb806ac239f26/tools/generator-terraform/generator/mappings/assignment_direct.go#L658-L665

i tried with LoadTestService 2022-12-01 the AccTest raise such error below because the Read method will always set the `encryption` field even the api return the `encryption` field as nil.

```
        Terraform used the selected providers to generate the following execution
        plan. Resource actions are indicated with the following symbols:
        -/+ destroy and then create replacement
        
        Terraform will perform the following actions:
        
          # azurerm_load_test.test must be replaced
        -/+ resource "azurerm_load_test" "test" {
              ~ data_plane_uri      = "8dfd843b-f118-48a4-8f6f-f68e7b02f708.westeurope.cnt-prod.loadtesting.azure.com" -> (known after apply)
              ~ id                  = "/subscriptions/xxxx/resourceGroups/acctestrg-xxxxxxx/providers/Microsoft.LoadTestService/loadTests/acctest-230508095009469840" -> (known after apply)
                name                = "acctest-230508095009469840"
                # (2 unchanged attributes hidden)
        
              - encryption { # forces replacement
                  - identity {}
                }
            }
        
        Plan: 1 to add, 0 to change, 1 to destroy.
--- FAIL: TestAccLoadTest_basic (157.30s)
```

### PR DESCRIPTION
**Pull request by wuxu92**
This PR is to fix some issues of the terraform generator tool.

1. rename schema field if there is a field with the same name exists when flattenning properties.
2. support transform legacy identity type when generate mapping functions.
3. should not set schema state if the field of sdk model is nil. resolves #2483
4. add model2model mappings of nestedItems. or generated resource will missing mapXXX functions.
5. add an acc test dependency management and fix acc test config generation.

### CODE REVIEW

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @wuxu92 

Thanks for this PR.

I've taken a look through and unfortunately this PR's going in quite a different direction to where we're intending to take the Terraform generator - so I'm not sure we're going to be able to proceed with this PR in it's current state.

Namely, we're intentionally avoiding adding workarounds within the codebase in favour of fixing the source data (aside from where it's expected that a PR will be merged in the short-term/to unblock us).

It's worth also calling out that a key design decision of Pandora is that all of the generated files can be removed at any time - and re-run (which is how the Terraform Generator works today) - so unfortunately allowing changes to the auto-generated files isn't something we can ship - since this'd lead to workarounds atop workarounds, which ultimately cause more issues long-term.

As you've identified, the root-cause of the issue with the Load Test Service resource is that the `identity` field is duplicated within one of the nested blocks, but is being top-levelled anyway and overwriting the original value. As such instead of taking the approaches outlined in this PR, I think this'd be best handled as a model processor - since we can validate that the field doesn't exist within the parent resource, and work from there (and validate that if it does, we're _intentionally_ ignoring/renaming the field for now).

Given the number of changes needed within this PR and that we're going to need to take a fundamentally different approach here - whilst I'd like to thank you for this contribution, I'm going to close this PR for the moment - but we'll take a look into the root-cause of this one in the not-too-distant future, since we're aware that we need to move off of this older API Version.

Thanks!

File: tools/generator-terraform/generator/resource/resource.go [hunk 1]
```diff
@@ -23,22 +24,41 @@ func Resource(input models.ResourceInput) error {
...
+// skip generate file if NOT CONTAINS 'manual changes will be overwritten' in file content
+func shouldSkipGenFile(file string) bool {
+	if f, err := os.Open(file); err == nil {
+		var buf = make([]byte, 512)
+		if _, err := f.Read(buf); err == nil {
+			return !bytes.Contains(buf, []byte("manual changes will be overwritten"))
+		}
+	}
+	return false
+}
```
[the remark this answers is not in the document]
Comment (wuxu92) [reply tools/generator-terraform/generator/resource/resource.go:64]: thanks. it looks reasonable to not skip specific files.

File: tools/importer-rest-api-specs/components/testattributes/test_dependency.go [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+	"fmt"
+	"strings"
+)
+
+type TestDep interface {
+	Name() string
+	GenConfig() string
+	RefValue(field ...string) string
+	Deps() []TestDep
+}
```
[the remark this answers is not in the document]
Comment (wuxu92) [reply tools/importer-rest-api-specs/components/testattributes/test_dependency.go:13]: for most of resources, the dependencies is complex and diferent from one to others. an example as when both resource A and B dependent on a `azurerm_key_vault_key` resource, they may need a different  configuraiton of the `azurerm_key_vault` resource to create the `key` resource (this is what LoadTestService required). is there a simple way to handle it?

**Comment by tombuildsstuff** [comment]
hey @wuxu92 

Thanks for pushing the changes to this PR - apologies for the delayed re-review here.

As mentioned in https://github.com/hashicorp/pandora/pull/2576, we've ended up reworking some of the internals about how the Terraform Resource Schema, Mappings and Test Generation works such that this PR has been superseded.

At this point in time we've looked to implement this in a slightly different manner, temporarily removing the `encryption` block (since there's a number of changes needed to support this) - which are being tracked in https://github.com/hashicorp/pandora/issues/2608 - and as such the Load Test resource has been updated to the latest API version as of https://github.com/hashicorp/pandora/pull/2609.

Since that's currently working it's way through into the Terraform Provider, since this PR was primarily to unlock the Load Test Service, whilst I'd like to thank you for this contribution I'm going to close this PR for the moment - however the remaining issues needed to pipe this additional field through are being tracked in #2608.

Thanks!

**Pull request closed without merging**